solr/core/src/java/org/apache/solr/handler/component/SearchHandler.java

```diff
@@ -500,6 +509,29 @@ public void handleRequestBody(SolrQueryRequest req, SolrQueryResponse rsp) throw
     }
   }
 
+  private void tagRequestWithRequestId(ResponseBuilder rb) {
+    String rid = getRequestId(rb.req);
+    if (StringUtils.isBlank(rb.req.getParams().get(CommonParams.REQUEST_ID))) {
+      ModifiableSolrParams params = new ModifiableSolrParams(rb.req.getParams());
+      params.add(CommonParams.REQUEST_ID, rid);//add rid to the request so that shards see it
+      rb.req.setParams(params);
+    }
+    if (rb.isDistrib) {
+      rb.rsp.addToLog(CommonParams.REQUEST_ID, rid); //to see it in the logs of the landing core
+    }
+  }
+
+  public static String getRequestId(SolrQueryRequest req) {
+    String rid = req.getParams().get(CommonParams.REQUEST_ID);
+    return StringUtils.isNotBlank(rid) ? rid : generateRid(req);
+  }
+
+  @SuppressForbidden(reason = "Need currentTimeMillis, only used for naming")
+  private static String generateRid(SolrQueryRequest req) {
+    String hostName = req.getCore().getCoreContainer().getHostName();
+    return hostName + "-" + req.getCore().getName() + "-" + System.currentTimeMillis() + "-" + ridCounter.getAndIncrement();
```

Comment:
Wow, this could be a long string, and it'll show up as a parameter and is thus kinda in-your-face when looking at logs.  And this, you propose, is by default. Isn't that a concern, albeit not huge?

Reply:
Does the string need to be human readable? A 32 bit hash based on node-id+counter would be unique enough?

Reply:
I'm a bit torn.

I agree that the length makes this a bit "in-your-face", and that a hash would be a good way to condense things.  I don't love the verbose-ness as things are now.  But, technically speaking, that loses us our uniqueness guarantee.  And if we're not shooting for strict-uniqueness here, then aren't we just as good with the NOW timestamp that was originally proposed but set aside for lack of uniqueness?

To zoom out, we've got three options for what the identifier is here:

1. **NOW timestamp** - concise with some semantic meaning, but not guaranteed to be unique
2. **RequestID** - very semantically meaningful and guaranteed to be unique, but "in your face"
3. **Hashed RequestID** - concise, but no semantic meaning and not guaranteed to be unique (though slightly more unique than a NOW timestamp?)

I originally implemented (1), and then moved to (2) based on feedback in JIRA.  But if there's no consensus there then maybe I'll go back to (1).  In any case I don't see much advantage in (3) over (1)

Reply:
Lmk if anyone has a strong opinion here.  I've gotten feedback from both directions (make it unique (but verbose) vs make it concise (but not unique)).  I'm going to just choose one in the next day or two to avoid dragging things out.

Reply:
I think the uniqueness goal is important (thus don't need hash), but we could cut back on the verbose-ness.  node ID + counter is, I think, about right.